eth/downloader/queue.go

```diff
@@ -473,60 +757,94 @@ func (q *queue) Deliver61(id string, blocks []*types.Block) (err error) {
 		q.hashQueue.Push(hash, float32(index))
 	}
 	// If none of the blocks were good, it's a stale delivery
-	if len(errs) != 0 {
-		if len(errs) == len(blocks) {
-			return errStaleDelivery
-		}
+	switch {
+	case len(errs) == 0:
+		return nil
+
+	case len(errs) == 1 && (errs[0] == errInvalidChain || errs[0] == errInvalidBlock):
+		return errs[0]
+
+	case len(errs) == len(request.Headers):
```

Comment:
Originally the errs array was collecting various types of errors, but by now all other errors have been handled separately (and checked for specifically in the above `case`s), so from this line onward `errs` can only contain errors related to getting data not requested. If all the data pieces were flagged as not requested, it's stale.